### ISSUE TITLE
Issues loading ppolicy module and generating debug logs

### ISSUE THREAD

**Issue by mwirtala** [osixia/docker-openldap#547]
I've seen other issues listed around loading ppolicy.ldif but my issue seems to begin where the proposed solutions for those issues end: I have a total of 5 custom .ldif files that I wish to seed my container with; the first 2 of these .ldif files seed users and groups and both work as expected. However, beginning with my 3rd .ldif, which by my understanding should load the ppolicy module prior to applying the overlay (done in my 4th ldif) and defining password policy values (my 5th ldif) I am having difficulty getting these ldif files to be processed as expected, and also having difficulty getting any additional info via logging.

I am running k8s and my openldap service is being managed via a helm chart. If I exec into my running openldap pod I do see all 5 of my custom .ldif files in /image/service/slapd/assets/config/bootstrap/ldif/custom. Furthermore, if I then manually run ldapmodify on the third ldif I am then able to successfully load the ppolicy module and then running ldapadd on the 4th and 5th ldifs works.

I have confirmed that in my pod CONTAINER_LOG_LEVEL has been set to 5 (trace) and see this in my pod logs as well. Howver, I do not see any of the expected debug level logging I'd expect from looking through the slapd startup.sh script (e.g. "Procesing file" lines for any custom ldif files. Furthermore, based on the logic around running ldapmodify when changetype is present in an ldif it seems like startup.sh should be running the appropriate command to succeed. Output of slapcat also confirms that ppolicy has not been loaded. Again, my first 2 .ldif files to seed users and groups both do get processed as expected.

Here is the contents of my 3rd ldif, where things start to fail:
dn: cn=module{0},cn=config
changeType: modify
add: olcModuleLoad
olcModuleLoad: ppolicy

So: Any ideas on (1) Why ppolicy may not be getting loaded and (2) Why setting the log level is not printing the expected debug log messages from startup.sh?

**Comment by bjozet** [comment]
Regarding `CONTAINER_LOG_LEVEL`: I suspect you're not supposed to actually pass that environment variable, but rather passing the `--loglevel trace` option to docker.

This did show `set -x` output of `startup.sh` for me when trying locally on my computer:
```
docker run -it osixia/openldap:1.5.0 bash --loglevel trace
``` 


Whereas this did not enable trace-logging:
```
docker run -it -e CONTAINER_LOG_LEVEL=5 osixia/openldap:1.5.0 bash
```

Not sure how you'd translate that into k8s/pod/helm terms though :)

**Comment by mwirtala** [comment]
Thanks much. Apologies for my lack of clarity; I'm actually controlling the log level in my helm chart by setting the logLevel attribute in my values.yaml file, e.g. "logLevel: trace" after which I see the CONTAINER_LOG_LEVEL value change in my kubectl logs <pod-id> output. On further inspection, it actually seems like I might be looking at the incorrect logs somehow, as I also don't see info level log-helper statements from startup.sh in the kubectl logs <pod-id> output. I'm pretty new to all of this (if that wasn't already apparent :)) so perhaps I just need to learn what log stream to look to to find the log-helper logs...

I'll try setting up a direct docker run command tomorrow with ldif inputs representative of my helm chart custom ldifs to see if I can get the docker run log output to figure out what might be happening without the k8s/helm layer...

**Comment by mwirtala** [comment]
Hi,

Apologies for my delay in commenting back here: An unexpected matter prevented me from following up on the direct docker run discussed above until now.
Further apologies for my lack of code formatting below: I'm visually impaired and reliant on a screen reader and so am sticking to raw text at present.

So: I set up a docker build of the docker-openldap image with my custom ldif files and, while in this context I now seem to be getting the ppolicy module loaded, I'm running into an error when trying to overlay the module in my subsequent ldif file.

Here's the content of that ldif file:

dn: olcOverlay=ppolicy,olcDatabase={1}hdb,cn=config
objectClass: olcOverlayConfig
objectClass: olcPPolicyConfig
olcOverlay: ppolicy
olcPPolicyDefault: cn=ppolicy,dc=mycompany

... and the error that I get with docker run is:

	matched DN: cn=config
adding new entry "olcOverlay=ppolicy,olcDatabase={1}hdb,cn=config"

ldap_add: Insufficient access (50)
adding new entry "olcOverlay=ppolicy,olcDatabase={1}hdb,cn=config"
***  ERROR  | 2021-03-23 19:54:48 | /container/run/startup/slapd failed with status 50

Any ideas here? Thank much!

**Comment by NoxInmortus** [comment]
Hello, idk if it can help you, but that's how I did it (with v1.4.0):
```
ldapadd -Y EXTERNAL -H ldapi:/// -f /etc/ldap/schema/ppolicy.ldif
ldapadd -x -w adminPassword -D "cn=admin,dc=MyDomain" -f oupolicy.ldif
ldapadd -Y EXTERNAL -H ldapi:/// -f ppmodule.ldif
ldapadd -Y EXTERNAL -H ldapi:/// -f ppolicyoverlay.ldif
ldapadd -x -w adminPassword -D "cn=admin,dc=MyDomain" -f passwordpolicy.ldif

# oupolicy.ldif
dn: ou=ppolicies,dc=MyDomain
objectClass: top
objectClass: organizationalUnit
ou: ppolicies
description: Password policy

# ppmodule.ldif
dn: cn=module{0},cn=config
changetype: modify
add: olcModuleLoad
olcModuleLoad: ppolicy

# ppolicyoverlay.ldif
dn: olcOverlay=ppolicy,olcDatabase={1}mdb,cn=config
objectClass: olcOverlayConfig
objectClass: olcPPolicyConfig
olcOverlay: ppolicy
olcPPolicyDefault: cn=default,ou=ppolicies,dc=MyDomain

# passwordpolicy.ldif
# Examples : https://www.zytrax.com/books/ldap/ch6/ppolicy.html
dn: cn=default,ou=ppolicies,dc=MyDomain
cn: default
objectClass: pwdPolicy
objectClass: pwdPolicyChecker
objectClass: device
objectClass: top
pwdAllowUserChange: TRUE
pwdAttribute: userPassword
pwdCheckQuality: 2
pwdCheckModule: pqchecker.so
pwdExpireWarning: 0
pwdFailureCountInterval: 0
pwdGraceAuthNLimit: 0
pwdInHistory: 0
pwdLockout: TRUE
pwdLockoutDuration: 7200
pwdMaxAge: 0
pwdMaxFailure: 5
pwdMinAge: 0
pwdMinLength: 8
pwdMustChange: FALSE
pwdSafeModify: FALSE
```

**Comment by mwirtala** [comment]
Thanks much - We are trying to have the adding of a default policy automated which is why I am trying to do this with some custom .ldif fragments added to the image/services/slapd/assets/config/bootstrap/ldif/custom/ directory, which by my understanding are being processed via the image/services/slapd/startup.sh script. Looking at that script, it seems like it should be attempting an ldapadd command equivalent to the one you suggest (on line 169 of my repo's version of startup.sh), which should attempt -Y external authentication first and then fall back to admin authentication if that command fails...

Apologies if I'm misunderstanding; are you adding this command somewhere else so that the ppolicy module is overlayed automatically as part of the docker build/run process?

**Comment by bjozet** [comment]
> adding new entry "olcOverlay=ppolicy,olcDatabase={1}hdb,cn=config"
> ldap_add: Insufficient access (50)

@mwirtala Not knowing how you execute your ldapadd or ldapmodify; could it be as simple as you not using the appropriate user? I haven't looked it up myself, but I could suspect that you need to authenticate as cn=admin,cn=config or whatever value the olcRootDN specifies in your setup, and not `cn=admin,dn=example,dn=com`?

I suspect that it might work for @NoxInmortus due to using `-Y EXTERNAL` with `ldapi` (ldap over ipc). Perhaps you could look into that.

**Comment by NoxInmortus** [comment]
Hello,

It's been a while since I worked on my openldap, I think I tried to use `image/services/slapd/assets/config/bootstrap/ldif/custom/` unsuccessfuly. It may depends on what you're trying to use, but as I saw many related reported issues I gave up on this method

- https://github.com/osixia/docker-openldap/issues/538
- https://github.com/osixia/docker-openldap/issues/449
- https://github.com/osixia/docker-openldap/issues/433
- https://github.com/osixia/docker-openldap/issues/402
- https://github.com/osixia/docker-openldap/issues/320
- And more...

To import my ldif files, I went in the container and executed the `ldap` commands I mentionned above. I don't recall everything I tried but there was a lot of different attempt to import the ppolicy module, and finally this was the only process that worked for me. Just keep in mind that I'm not an OpenLDAP expert at all so my sayings may not be optimal. But it works.

**Comment by mwirtala** [comment]
Hi, thanks so much for all the help here! I actually figured something out on the openldap side: In my docker-openldap repo it seems like I should have been specifying the mdb OLCDatabase rather than the hdb. Sorry about this; the helm chart I am still trying to get to work properly uses a version with hdb. Anyways, switching my overlay ldif from hdb to mdb allowed me to successfully apply the ldifs automatically by placing them in image/services/slapd/assets/config/bootstrap/ldif/custom/.

So: It seems my issue may actually be on the helm side of things. I'm going to see if I can update the container version used by the helm chart to match the version I ran directly from docker just now and may open up an issue in the helm repo if needed. But I am closing this one out since things on the docker-openldap side do seem to work for me. Thanks again!
